ctp2_code/ai/strategy/scheduler/scheduler.cpp

```diff
@@ -867,6 +883,231 @@ void Scheduler::Match_Resources(const bool move_armies)
 		}
 	}
 
+	if (needToGoBackOne)
+	{
+		// by deleting goals we skipped the last goal - go back and do it.
+		Goal_List::iterator goal_iter = m_goals.end();
+		--goal_iter;
+		needToGoBackOne = false;
+		if (committed_agents >= total_agents)
+		{
+			Assert(committed_agents == total_agents);
+		}
+
+		Goal_ptr goal_ptr = static_cast<Goal_ptr>(*goal_iter);
+		Utility     oldMatchValue = goal_ptr->Get_Matching_Value();
+
+#if defined(_DEBUG) || defined(USE_LOGGING)
+		MapPoint pos = (goal_ptr->Get_Target_Army().m_id == 0 || goal_ptr->Get_Target_Army().IsValid()) ? goal_ptr->Get_Target_Pos() : MapPoint(-1, -1);
+		AI_DPRINTF(k_DBG_SCHEDULER, m_playerId, goal_ptr->Get_Goal_Type(), -1, ("\n"));
+		AI_DPRINTF(k_DBG_SCHEDULER, m_playerId, goal_ptr->Get_Goal_Type(), -1, ("\n"));
+		AI_DPRINTF(k_DBG_SCHEDULER, m_playerId, goal_ptr->Get_Goal_Type(), -1,
+			("[%d] Starting to match resources to %s: %x (x=%d,y=%d), match %d, %s\n",
+				count, g_theGoalDB->Get(goal_ptr->Get_Goal_Type())->GetNameText(), goal_ptr, pos.x, pos.y, oldMatchValue, (g_theWorld->HasCity(pos) ? g_theWorld->GetCity(pos).GetName() : "field")));
+		count++;
+#endif
+
+		if (oldMatchValue == Goal::BAD_UTILITY)
+		{
+			AI_DPRINTF(k_DBG_SCHEDULER, m_playerId, goal_ptr->Get_Goal_Type(), -1,
+				("\t\tGOAL (goal: %x) -- First goal with bad utility - stop matching\n",
+					goal_ptr));
+
+			AI_DPRINTF(k_DBG_SCHEDULER, m_playerId, goal_ptr->Get_Goal_Type(), -1, ("\n"));
+			AI_DPRINTF(k_DBG_SCHEDULER, m_playerId, goal_ptr->Get_Goal_Type(), -1, ("\n"));
+			// Assuming that the list is still sorted,
+			// and the following has only Goal::BAD_UTILITY
+		}
+
+		Utility newMatchValue = goal_ptr->Compute_Matching_Value(false);
+
+		if (newMatchValue == Goal::BAD_UTILITY)
+		{
+			AI_DPRINTF(k_DBG_SCHEDULER, m_playerId, goal_ptr->Get_Goal_Type(), -1,
+				("\t\tGOAL (goal: %x)(agent count: %d) -- Goal with bad utility, trying agian in next cycle.\n",
+					goal_ptr, goal_ptr->Get_Agent_Count()));
+
+			//		Assert(goal_ptr->Get_Agent_Count() == 0); // Is still ok
+					// City garrison problem
+			goal_ptr->Rollback_All_Agents(); // Just roll back but don't report to the build list
+
+			// Actually should be checked in the next cycle, but there still seems to be something wrong.
+			Goal_List::iterator tmp_goal_iter = goal_iter;
+			m_goals.erase(tmp_goal_iter);
+
+			/*
+			// Move to the end
```

Comment:
Please do not add 'new' commented code. We have our source-code repository which we can inspect when we need this code again. In the future it will be hard to understand why this code is commented out or will just make the reading harder/code dirty.

Reply:
This was copy+pasted from the existing code - I will remove this and the original when I improve this implementation.